### ISSUE TITLE
Incorrect results when using "Show items from unimported namespaces" to complete generic types

### ISSUE THREAD

**Issue by davkean** [dotnet/roslyn#36730]
**Steps to Reproduce**:

I was trying to complete System.Collections.Generics.List<T>, but I got the non-generic version of this type:

![TypeCompletionForGenerics](https://user-images.githubusercontent.com/1103906/60062863-3d140700-973e-11e9-9df7-4f7c445121bf.gif)

**Expected Behavior**:
List<T> to be selected.

**Actual Behavior**:
List was selected.

**Comment by davkean** [comment]
tag @genlu

**Comment by davkean** [comment]
I played around with this with the normal completion, looks like it doesn't need to handle this situation, and this completion works as expected if the generic and non-generic live in the same namespace.

**Comment by genlu** [comment]
Yea, we have discussed this (and I totally forgot to file an issue). One potential solution we came up with is to keep track which namespaces/types are used in current project, and give that higher priority than others with same name.

**Comment by genlu** [comment]
Also, "`List` was committed" is the expected behavior here, since that's what you have selected from the list (or more accurately, selected for you). The actual problem is `List<T>` is not the selected one.

**Comment by davkean** [comment]
I think you need to rethink the design for generics in the completion window. In this world, it's now required that you know how generic types parameters are needed before you add "a using".

**Comment by CyrusNajmabadi** [comment]
> it's now required that you know how generic types parameters are needed before you add "a using".

I don't think this is surprising.  If there are many lists in the world (regardless of whether they are generic or not), you'd have to pick the right one you wanted to get hte right import added.  Wouldn't it make sense that you get the list, and you have to select the one from `System.Collections.Generic` in order to get that added?

Or am i missing something.

**Comment by davkean** [comment]
The impact is that with this feature turned on its now become impossible to type `List<` without List being selected automatically and committed.

**Comment by drewnoakes** [comment]
@davkean what were your keystrokes in the screen recording?

**Comment by davkean** [comment]
I typed 'L', 'I', 'S', 'T', '<'.

**Comment by CyrusNajmabadi** [comment]
Ah, i see the concern Dave has.  Note: i don't think this is specific to Lists or generic.  If you type anything that now matches an importable type, and you hit a commit character, you will get the import added.  Right?  So even just typing something undefined, but hitting .dot, you could complete out to something unintended.

Should then perhaps be soft-selected?  That way they never commit, except through explicit user action?

**Comment by mhutch** [comment]
FWIW we handled this problem in MonoDevelop by suppressing commit when the typed character caused a better filter match, even if it was a commit character.

**Comment by davkean** [comment]
@genlu ^ This sounds like a great idea.

Does it dimiss the completion window, or keep completing?

**Comment by mhutch** [comment]
It continues completing, so you can filter down to any specific type.

**Comment by genlu** [comment]
Thank @mhutch, I like this idea too. Will look into this.
